src/fread.c

```diff
@@ -1534,10 +1531,20 @@ int freadMain(freadMainArgs _args) {
         // field) since we rely on that logic to avoid the copy below when fileSize$4096==0 but there is a final eol ok.
         // TODO: portable way to discover relevant page size. 4096 is lowest common denominator, though, and should suffice.
       } else {
-        const char *msg = _("This file is very unusual: it ends abruptly without a final newline, and also its size is a multiple of 4096 bytes. Please properly end the last row with a newline using for example 'echo >> file' to avoid this ");
-        if (verbose) DTPRINT(_("  File ends abruptly with '%c'. Copying file in RAM. %s copy.\n"), eof[-1], msg);
+        const char *msg = _("This file is very unusual: it ends abruptly without a final newline, and also its size is a multiple of 4096 bytes. Please properly end the last row with a newline using for example 'echo >> file'");
+        if (verbose)
+          DTPRINT(_("  File ends abruptly with '%c'. Final end-of-line is missing. Copying file in RAM. %s.\n"), eof[-1], msg);
         // In future, we may discover a way to mmap fileSize+1 on all OS when fileSize%4096==0, reliably. If and when, this clause can be updated with no code impact elsewhere.
-        copyFile(fileSize, msg, verbose);
+        double time_taken = copyFile(fileSize);
+        if (time_taken < 0) {
```

Comment:
In your experience as a professional software developer, does it pay to account for very unlikely corner cases, such as the system clock being very coincidentally adjusted backwards right as `copyFile` is running? If you think it's too unlikely to care about, that's probably fine.

Maybe we could return `NaN` on error, but that's a weird error return code.

Reply:
I was vaguely musing if that's possible as I wrote this.

I think we can wait for a user report before fussing here

copyFile should almost never be called. the two cases are (a) one-column files with multiple trailing missing rows and (b) files without trailing newline, in both cases only if the file is an exact multiple of 4096 bytes.

for all that to be true AND the clock to run backwards, it feels like the user is just messing with us :)

Reply:
one easy way to make this even more impossible to trigger by mistake would be to only check copy file==-1.0 instead of just being negative. then the user would somehow have to have all of the above with an impossibly exact regression of the system time.

Reply:
If someone manages to trigger exactly that, well, it would be too late to buy lottery tickets.